src/serializer/ResidualHeapVisitor.js

```diff
@@ -582,98 +589,131 @@ export class ResidualHeapVisitor {
+        let binding: PropertyBinding = ((propertyBinding: any): PropertyBinding);
+        let object = binding.object;
+        if (object instanceof ObjectValue && createdObjects.has(object)) continue; // Created Object's binding
+        if (object.refuseSerialization) continue; // modification to internal state
+        if (object.intrinsicName === "global") continue; // Avoid double-counting
+        this.visitObjectProperty(binding);
+      }
+      // Handing of ModifiedBindings
+      for (let [additionalBinding, previousValue] of modifiedBindings) {
+        let modifiedBinding = additionalBinding;
+        let residualBinding;
+        this._withScope(functionValue, () => {
+          // Also visit the original value of the binding
+          residualBinding = this.visitBinding(functionValue, modifiedBinding.name);
+          invariant(residualBinding !== undefined);
+          // Fixup the binding to have the correct value
+          // No previousValue means this is a binding for a nested function
+          if (previousValue) residualBinding.value = this.visitEquivalentValue(previousValue);
+          invariant(functionInfo !== undefined);
+          if (functionInfo.modified.has(modifiedBinding.name)) residualBinding.modified;
+        });
+        invariant(residualBinding !== undefined);
+        invariant(funcInstance !== undefined);
+        funcInstance.residualFunctionBindings.set(modifiedBinding.name, residualBinding);
+        // Only visit it if there is already a binding (no binding means that
+        // the additional function created the binding)
+        //if (residualBinding && modifiedBinding.value !== residualBinding.value) {
+        //if (modifiedBinding.value !== previousValue) {
+        let newValue = modifiedBinding.value;
+        invariant(newValue);
+        this.visitValue(newValue);
+        residualBinding.modified = true;
+        // This should be enforced by checkThatFunctionsAreIndependent
+        invariant(
+          !residualBinding.additionalFunctionOverridesValue,
+          "We should only have one additional function value modifying any given residual binding"
+        );
+        if (previousValue) residualBinding.additionalFunctionOverridesValue = functionValue;
+        modifiedBindingInfo.set(modifiedBinding, residualBinding);
+        //}
+      }
+      invariant(result instanceof Value);
+      if (!(result instanceof UndefinedValue)) this.visitValue(result);
+    };
+    this.additionalFunctionValueInfos.set(functionValue, {
+      functionValue,
+      captures: functionInfo.unbound,
+      modifiedBindings: modifiedBindingInfo,
+      instance: funcInstance,
+    });
+    this.visitGenerator(generator);
+    this._withScope(generator, visitPropertiesAndBindings);
+    // Re-visit any bindings corresponding to unbound values to make sure that
+    // they're serialized in the correct scope
+    parentScopeVisitQueue.push(() => {
```

Comment:
Not sure what the exact reason for this `parentScopeVisitQueue` thing is. Instead, could you pull the `this.realm.evaluateAndRevertInGlobalEnv(` inside of this function, and then run the code you pass to `parentScopeVisitQueue` at the very end of this function?

Reply:
I could, I just thought it might be a little cleaner to not have a giant anonymous function.